### ISSUE TITLE
Bug in PTFF causing wait 9064 for z/OS 2.4 under z/VM 7.1

### ISSUE THREAD

**Issue by Fish-Git** [SDL-Hercules-390/hyperion#363]
Ever since [**Miscellaneous-Instruction-Extensions Facility 2** and **3** support was committed](https://github.com/SDL-Hercules-390/hyperion/commit/f54122e2eb5018bb07ec18fe1553797e80fc6e84) (Facilities 58 and 61 respectively), z/OS 2.4 now no longer IPLs under z/VM 7.1:

<pre>
  ...[snipped]...
  CSV410I APF FORMAT IS NOW DYNAMIC                                     
  IEA598I TIME ZONE = W.06.00.00                                        
  <b>IEA304W SYSTEM WAIT STATE CODE 80009064 DURING IEAVNP21 INITIALIZATION</b>
</pre>

Manual SA38-0665-30 "MVS System Codes" describes 9064 as:

> _**064**_
>
> _**Explanation:** During nucleus initialization program (NIP) processing, NIP detected a software or hardware error. An attempt was made to enter the recovery termination manager (RTM). This was trapped by NIP because the system is in a variable state of initialization and, in particular, RTM is not fully initialized._

> _**009** A program check occurred. Accompanying message IEA304W further explains this wait state and entry code. If the message does not appear on the console, you can find the message in the wait state message area (WSMA). The WSMA is described in z/OS MVS Data Areas..._

> _**System programmer response:** &nbsp; ......_
> * _For table entry code X'009', the program check OLD PSW indicates the type of exception and where it occurred._

Hercules itself was not reporting any type of program check as having occurred (even when using `OSTAILOR DEFAULT`), so when the disabled wait occurred (dropping my z/OS VM guest into a `CP` prompt), I did:

```
00: d t150.10
00: R00000150  04043000 80000000 00000000 1F902AEE 0E *................*
```

Then, trying again with `trace instruction pswa 1F902AC0-1F902AF0 run` active, I discovered that it was the PTFF instruction (opcode x'0104') that was blowing up with a Specification Exception. (The PTFF instruction, being semi-privileged, is always intercepted and simulated by z/VM when running in SIE mode.)

So I tried one last time using `trace instruction pswa 1F902AC0-1F902AF0 run cmd display gg0-1` and discovered PTFF instruction's GR0 function code was "5"!

Looking at the Hercules code for the PTFF instruction ([`esame.c` line 5392](https://github.com/SDL-Hercules-390/hyperion/blob/135ac68c35bbdd80ab9c7c8aa3b1b8b275550eff/esame.c#L5392)), we can see that support for function code 5 is not defined! Hercules only has function codes 0-3 and 64-67 defined (see [`esa390.h` lines 2480-2487](https://github.com/SDL-Hercules-390/hyperion/blob/135ac68c35bbdd80ab9c7c8aa3b1b8b275550eff/esa390.h#L2480-L2487
)), but not function code 5!

However, checking the current version of the "zArchitecture Principles of Operation" manual (SA22-7832-12), we can clearly see that the PTFF instruction does indeed define function code 5, as well as many others which are not defined in Hercules!

What's going on?!

I spent some time researching this and it appears, based on Bernard's comments (who, according to "git blame", was the one that first added the instruction back in 2005), it was in the SA22-7832-04 Principles of Operation manual (released in Sept. 2005) that the PTFF instruction was first described (as part of the (then) new "TOD-Clock-Steering Facility", bit 28), and back then there were indeed only 8 function codes defined: 0-3 and 64-67.

But then additional PTFF function codes were added (4 and 5) and others dropped (64, 66 and 67) sometime in March 2015 when SA22-7832-10 was released, and Hercules code was unfortunately never updated. &nbsp;`:(`

And then finally, around Sept. 2017 when SA22-7832-11 was released (describing the new "Multiple-Epoch Facility", bit 139), additional function codes were defined (10, 13, 73 and 77), which is where we are today. (SA22-7832-12, the current Principles of Operation manual as of today, does not document any change to the PTFF instruction not already described in SA22-7832-11).

&nbsp;
Sooo....
&nbsp;

We need to, _at the minimum_, add the missing support for function codes 4, 5 and 69, _BUT..._

&nbsp;
**I'm more concerned about what to do about our existing support for function codes 64, 66 and 67! Should we _drop_ support for those codes? I'm somewhat reluctant to do so since they _did_ exist for almost 10 years (from 2005 to 2015) and thus there might still be some operating systems out there that would then break if we removed support for those function codes, _BUT..._**
&nbsp;

The current Principles of Operation _does_ document them as _no longer being defined_, and therefore _should_ technically cause a Specification Exception or Condition Code 3 **`(*)`** when specified in the instruction:

> _**Special Conditions**_
> 
> A specification exception is recognized and no other action is taken if any of the following occurs:
> 
> 1.  Bit 56 of general register 0 is not zero.
> 2.  Bits 57-63 of general register 0 specify an unassigned or uninstalled function code.

> _**Resulting Condition Code:**_
>
> 0 Requested function performed
> 1 -- 
> 2 -- 
> 3 Requested function not available 

> _**Priority of Execution:**_
> ...
> 8.  Specification exception due to bit 56 of general register 0 being one or bits 57-63 of general register 0 specifying an unassigned or uninstalled function code.
> 9.  Privileged-operation exception due to attempt to issue PTFF control function in the problem state.
> 10. Condition code 3 due to a control function not available in supervisor state. 
> 11. Access exceptions for an access to the parameter block.
> ...

&nbsp;
So how should we handle this? I'm thinking the only way we can possibly support both is to maybe define some new Hercules Facility that, if enabled, causes PTFF to behave the way it originally did (i.e. the way Hercules currently has it erroneously implemented), with the default being disabled, thereby causing it to behave the way it _SHOULD_ behave according to today's SA22-7832-12 Principles of Operation manual by default.

Feedback appreciated!

Thanks!
___
**`(*)`** The cc3 vs. Specification Exception issue: Please let me know if you agree with me that, if I'm reading the manual correctly, cc3 should only be returned if the instruction is being executed in Supervisor State, whereas if it is executed in Problem State, a Specification Exception should be thrown instead.

I'm going to _TRY_ coding the missing function codes (or at least the missing function code 5 at the very minimum!), but I've only just started on this project so I don't know how difficult it will eventually prove to be.

One thing is for _darn_ sure though:

1. Our current implementation is clearly _wrong!_
2. I'm definitely _**not**_ going to mess with trying to implement the new Multiple-Epoch Facility function codes until such time (heh!) as the Multiple-Epoch Facility _itself_ is coded. (We don't currently support the Multiple-Epoch Facility yet.)
___
_p.s. It still pisses me off that defining/implementing something as innocuous as the Miscellaneous-Instruction-Extensions Facility would somehow trigger z/VM to simulate the PTFF instruction so very differently as to cause a guest to not be able to IPL as a result! I mean, what the FRICK does Miscellaneous-Instruction-Extensions Facility have to do with the TOD-Clock-Steering Facility?! NOTHING! (Sheesh!)_

**Comment by arfineman** [comment]
Hi, Fish,
Try turning on Bit 55 with your new Facility enhancements and see if that resolves your problem.
Best regards,

**Comment by Fish-Git** [comment]
> Try turning on Bit 55 with your new Facility enhancements and see if that resolves your problem.

How will that fix the problem? What is bit 55? According to the available documentation, facility bit 55 is undefined (or rather, defined as "Assigned to IBM Internal Use"). Do you know something we don't? What does bit 55 do? And more importantly, how will setting it magically fix the obvious bug I described? Have you tried it?

**Comment by arfineman** [comment]
zVM is very meticulous as to only simulate the underlying hardware. PTFF function code 05 is only available on a z13 or higher.
I'm guessing that the zOS is seeing the new facility bits and thinks its running on a z15 and issues the PTFF function 05. 
But zVM doesn't see z13 functionality and rejects function 05 as invalid. Bit 55 tells zVM it has z13 functionality.
I can't build Hercules from source. So I won't be able to test the new functions until a new release comes out. 
Best regards,

**Comment by Fish-Git** [comment]
Then based on what you've just said, setting bit 55 is _not_ going to help any:

> 1. PTFF function code 05 is only available on a z13 or higher.
> 2. Bit 55 tells zVM it has z13 functionality.

3. Therefore if bit 55 is set, z/VM will think it's running on a z13 and thus issue PTFF function code 05 as a result (which is only cause the same thing to happen that is already happening.)

Besides, if <span>#</span>1 is true (fc 05 only available on z13) and <span>#</span>2 is _not_ true (i.e. bit 55 is _off_, telling z/VM that z13 functionality is _not_ available), _**AND**_ `CPUMODEL $(EC12)` is specified in the Hercules configuration file (i.e. CPU model 2827, i.e. a model that _pre-dates_ the z13), THEN (if your claim is true) the problem should _not_ occur. Yes? Correct?

Well I have news for you Aaron: I already tried that and it _still_ fails in the exact same way.

Setting bit 55 is not going to accomplish anything. It's not going to magically fix the bug. The bug is that support for function code 05 is completely missing in Hercules, and setting bit 55 is not going to somehow make the missing support magically appear in Hercules (i.e. it's not going to magically cause the missing code to get written).

Nor is it going to prevent z/VM from issuing PTFF function code 05 either for that matter, based on what you've written.

I'm sorry Aaron, but your (i.e. the Hercules user's) only choice in this matter is to either:

1. Disable Facility 58.
2. Wait for the Hercules developers (i.e. _me!_) to add the missing support for function code 05 (as well as support for any additional function codes that might also be needed/required).

<span>#</span>1 is already known to work (i.e. is known to be a valid viable workaround/circumvention).

<span>#</span>2 is currently in progress. I've already finished with coding it (only a few minutes ago as a matter of fact), but have not yet had time to test it. I've been up all night and need some much needed sleep!

**Comment by arfineman** [comment]
Hi Fish,
My understating was that the scenario you described, was working before and it only failed after you added facility 58 and 61. 
That let me to believe that zOS is only issuing the function 05, thinking it is running on a z15 and ignoring the CPU model.
Best regards,

**Comment by Fish-Git** [comment]
> My understating was that the scenario you described, was working before and it only failed after you added facility 58 and 61.

That is true, yes. The problem only started occurring when facility 58 was added. But it was always occurring _only_ when facility 58 was enabled. It didn't matter what the CPU mode was. It happened _regardless_ of what CPU model was specified. The trigger for the problem was apparently the enabling of facility 58.

> That let me to believe that z/OS is only issuing the function 05, thinking it is running on a z15 and ignoring the CPU model.

That is apparently also true.

But setting bit 55, which if I'm understanding you correctly, tells z/OS (and presumably z/VM too) that it's running on a z13, won't help since it is the very fact that z/OS (and z/VM too presumably) thinks that it is running on a z13 or greater that is causing the problem. That is to say, the problem only occurs whenever z/OS (and z/VM) think they are running on a z13 or greater (thereby causing then to issue function code 05), and setting bit 55 _doesn't change that._

Whether bit 55 is zero, or whether bit 55 is one (either way), z/OS and z/VM are _still_ going to believe they are running on a z13 and end up issuing function code 05. It is apparently facility bit _58_ that is causing z/OS and z/VM to believe they are running on a z13, not bit 55.

So as I said, setting bit 55 is _not_ going to help matters any. The only way this problem is going to be fixed is by adding the missing function code 05 support to Hercules (which I have just finished testing and seems to work just fine and which I will be committing the fix for shortly).
